sync/syncer.go

```diff
@@ -199,98 +214,190 @@ func (s *Syncer) getLayerFromNeighbors(currenSyncLayer types.LayerID) (*types.La
 		return nil, err
 	}
 
-	blocksArr, err := s.fetchFullBlocks(blockIds)
-	if err != nil {
-		s.Error("could not get layer blocks %v", currenSyncLayer, err)
+	blocksArr := s.GetFullBlocks(blockIds)
+	if len(blocksArr) == 0 {
+		s.Error("could not any blocks  for layer  %v", currenSyncLayer)
 		return nil, err
 	}
 
 	return types.NewExistingLayer(types.LayerID(currenSyncLayer), blocksArr), nil
 }
 
-func (s *Syncer) fetchFullBlocks(blockIds []types.BlockID) ([]*types.Block, error) {
-	output := s.fetchWithFactory(BlockReqFactory(blockIds), s.Concurrency)
+func (s *Syncer) GetFullBlocks(blockIds []types.BlockID) []*types.Block {
+	output := s.fetchWithFactory(NewBlockWorker(s, s.Concurrency, BlockReqFactory(), blockSliceToChan(blockIds)))
 	blocksArr := make([]*types.Block, 0, len(blockIds))
 	for out := range output {
 		block := out.(*types.Block)
+		if err := s.ConfirmBlockValidity(block); err != nil {
+			s.Error("failed derefrencing block data %v %v", block.ID(), err)
+			continue
+		}
 
-		eligible, err := s.BlockEligible(&block.BlockHeader)
+		//data availability
+		txs, atxs, err := s.DataAvailabilty(block)
 		if err != nil {
-			s.Warning(fmt.Sprintf("failed checking eligiblety %v", block.ID()), err)
+			s.Error("data availabilty failed for block %v", block.ID())
 			continue
 		}
-		if !eligible {
-			s.Warning(fmt.Sprintf("block %v not eligible", block.ID()), err)
+
+		//validate blocks view
+		if valid := s.ValidateView(block); valid == false {
+			s.Error("block %v not syntacticly valid", block.ID())
 			continue
 		}
 
-		if err := s.syncMissingContent(block); err != nil {
-			s.Error(fmt.Sprintf("failed derefrencing block data %v", block.ID()), err)
+		if err := s.AddBlockWithTxs(block, txs, atxs); err != nil {
+			s.Error("failed to add block %v to database %v", block.ID(), err)
 			continue
 		}
 
-		s.Info("added block to layer %v", block.Layer())
+		s.Info("added block %v to layer %v", block.ID(), block.Layer())
 		blocksArr = append(blocksArr, block)
+	}
 
+	s.Info("done getting full blocks")
+	return blocksArr
+}
+
+func (s *Syncer) ConfirmBlockValidity(blk *types.Block) error {
+	s.Info("validate block %v ", blk.ID())
+
+	//check block signature and is identity active
+	pubKey, err := ed25519.ExtractPublicKey(blk.Bytes(), blk.Sig())
+	if err != nil {
+		return errors.New(fmt.Sprintf("could not extract block %v public key %v ", blk.ID(), err))
 	}
 
-	return blocksArr, nil
+	active, atxid, err := s.IsIdentityActive(signing.NewPublicKey(pubKey).String(), blk.Layer())
+	if err != nil {
+		return errors.New(fmt.Sprintf("error while checking IsIdentityActive for %v %v ", blk.ID(), err))
+	}
+
+	if !active {
+		return errors.New(fmt.Sprintf("block %v identity activation check failed ", blk.ID()))
+	}
+
+	if atxid != blk.ATXID {
+		return errors.New(fmt.Sprintf("wrong associated atx got %v expected %v ", blk.ATXID.ShortId(), atxid.ShortId()))
```

Comment:
@noamnelke  @zalmen this is failing the apptest on my branch 
"wrong associated atx got 056a0 expected 00000 "
 is the error im getting, it happens in layer 2 so i guess it should b 00000 ?

Reply:
There should be an IsGenesis() check, similar to the one in IsIdentityActive().
I am not sure if it should be inside this function, or before calling it.